File: backend/agent/tools/process.py
from __future__ import annotations

import asyncio
import os
import re
import subprocess
from typing import Any

from agent.tools.base import Tool


def _is_windows() -> bool:
    return os.name == "nt"


class ProcessListTool(Tool):
    """List running processes. Structured output (name, pid, optional session id)."""

    name = "process_list"
    description = (
        "List running processes. Returns a compact table: PID, name, optional session id. "
        "Filter by name (case-insensitive substring). Output is capped at 200 rows."
    )
    params = {
        "filter": {
            "type": "string",
            "description": "Substring to filter by process name (optional)",
        },
    }

    async def execute(self, args: dict[str, Any]) -> str:
        name_filter = str(args.get("filter", "")).strip().lower()
        try:
            if _is_windows():
                rows = await self._list_windows(name_filter)
            else:
                rows = await self._list_unix(name_filter)
        except subprocess.TimeoutExpired:
            return "error: timeout listing processes"
        except Exception as e:
            return f"error: {e}"

        if not rows:
            return "no processes" if not name_filter else f"no processes matching '{name_filter}'"
        out = "\n".join(f"{pid:>7}  {name}" for pid, name in rows[:200])
        if len(rows) > 200:
            out += f"\n... and {len(rows) - 200} more"
        return out

    async def _list_windows(self, name_filter: str) -> list[tuple[str, str]]:
        # tasklist /FO CSV /NH gives: "name","pid","session","sessionNum","mem"
        proc = await asyncio.to_thread(
            subprocess.run,
            ["tasklist", "/FO", "CSV", "/NH"],
            capture_output=True,
            text=True,
            timeout=15,
            encoding="utf-8",
            errors="replace",
        )
        rows = []
        for line in (proc.stdout or "").splitlines():
            parts = [p.strip().strip('"') for p in line.split('","')]
            if len(parts) < 2 or not parts[1].isdigit():
                continue
            pid, name = parts[1], parts[0]
            if name_filter and name_filter not in name.lower():
                continue
            rows.append((pid, name))
        return rows

    async def _list_unix(self, name_filter: str) -> list[tuple[str, str]]:
        # `ps -eo pid=,comm=` is portable and predictable.
        proc = await asyncio.to_thread(
            subprocess.run,
            ["ps", "-eo", "pid=,comm="],
            capture_output=True,
            text=True,
            timeout=15,
            encoding="utf-8",
            errors="replace",
        )
        rows = []
        for line in (proc.stdout or "").splitlines():
            m = re.match(r"^\s*(\d+)\s+(.+)$", line)
            if not m:
                continue
            pid, name = m.group(1), m.group(2).strip()
            if name_filter and name_filter not in name.lower():
                continue
            rows.append((pid, name))
        return rows


class ProcessKillTool(Tool):
    """Terminate a process by PID. Refuses obviously dangerous PIDs."""

    name = "process_kill"
    description = (
        "Terminate a process by PID. On Windows uses taskkill /F; on Unix uses kill -9. "
        "Always force-kills (no graceful shutdown). Refuses PIDs 0, 1, and the current process."
    )
    params = {
        "pid": {"type": "integer", "description": "Process ID to terminate"},
    }

    def is_dangerous(self, args: dict[str, Any]) -> bool:
        return True

    async def execute(self, args: dict[str, Any]) -> str:
        pid = args.get("pid")
        try:
            pid = int(pid)
        except (TypeError, ValueError):
            return "error: pid must be an integer"
        if pid <= 1:
            return "error: refusing to kill PID 0 or 1 (init/system)"
        if pid == os.getpid():
            return "error: refusing to kill self"

        try:
            if _is_windows():
                proc = await asyncio.to_thread(
                    subprocess.run,
                    ["taskkill", "/F", "/PID", str(pid)],
                    capture_output=True,
                    text=True,
                    timeout=10,
                    encoding="utf-8",
                    errors="replace",
                )
            else:
                proc = await asyncio.to_thread(
                    subprocess.run,
                    ["kill", "-9", str(pid)],
                    capture_output=True,
                    text=True,
                    timeout=10,
                    encoding="utf-8",
                    errors="replace",
                )
        except subprocess.TimeoutExpired:
            return f"error: timeout killing PID {pid}"
        except Exception as e:
            return f"error: {e}"

        if proc.returncode == 0:
            return f"killed PID {pid}"
        err = (proc.stderr or proc.stdout or "").strip() or f"exit {proc.returncode}"
        return f"error: failed to kill PID {pid}: {err[:200]}"

File: agent/tools/process.go
package tools

import (
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

type ProcessListTool struct{}

func (t *ProcessListTool) Name() string { return "process_list" }

func (t *ProcessListTool) Execute(args map[string]any) Result {
	filter := strings.ToLower(strings.TrimSpace(getStr(args, "filter")))
	var rows []string
	var err error
	if runtime.GOOS == "windows" {
		rows, err = listWindows(filter)
	} else {
		rows, err = listUnix(filter)
	}
	if err != nil {
		return R("error: " + err.Error())
	}
	if len(rows) == 0 {
		if filter != "" {
			return R("no processes matching '" + filter + "'")
		}
		return R("no processes")
	}
	var b strings.Builder
	for i, r := range rows {
		if i >= 200 {
			b.WriteString(fmt.Sprintf("\n... and %d more", len(rows)-200))
			break
		}
		b.WriteString(r)
		b.WriteString("\n")
	}
	return R(strings.TrimRight(b.String(), "\n"))
}

func listWindows(filter string) ([]string, error) {
	cmd := exec.Command("tasklist", "/FO", "CSV", "/NH")
	out, err := cmd.Output()
	if err != nil {
		return nil, err
	}
	var rows []string
	for _, line := range strings.Split(string(out), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		// CSV: "name","pid",...
		parts := strings.Split(line, "\",\"")
		if len(parts) < 2 {
			continue
		}
		name := strings.Trim(parts[0], "\" ")
		pidStr := strings.Trim(parts[1], "\" ")
		if _, err := strconv.Atoi(pidStr); err != nil {
			continue
		}
		if filter != "" && !strings.Contains(strings.ToLower(name), filter) {
			continue
		}
		rows = append(rows, fmt.Sprintf("%7s  %s", pidStr, name))
	}
	return rows, nil
}

var psRe = regexp.MustCompile(`^\s*(\d+)\s+(.+)$`)

func listUnix(filter string) ([]string, error) {
	cmd := exec.Command("ps", "-eo", "pid=,comm=")
	out, err := cmd.Output()
	if err != nil {
		return nil, err
	}
	var rows []string
	for _, line := range strings.Split(string(out), "\n") {
		m := psRe.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		pid, name := m[1], strings.TrimSpace(m[2])
		if filter != "" && !strings.Contains(strings.ToLower(name), filter) {
			continue
		}
		rows = append(rows, fmt.Sprintf("%7s  %s", pid, name))
	}
	return rows, nil
}

type ProcessKillTool struct{}

func (t *ProcessKillTool) Name() string { return "process_kill" }

func (t *ProcessKillTool) Execute(args map[string]any) Result {
	if _, ok := args["pid"]; !ok {
		return R("error: pid is required")
	}
	pid := int(getFloat(args, "pid"))
	if pid <= 1 {
		return R("error: refusing to kill PID 0 or 1 (init/system)")
	}
	if pid == os.Getpid() {
		return R("error: refusing to kill self")
	}
	var cmd *exec.Cmd
	if runtime.GOOS == "windows" {
		cmd = exec.Command("taskkill", "/F", "/PID", strconv.Itoa(pid))
	} else {
		cmd = exec.Command("kill", "-9", strconv.Itoa(pid))
	}
	out, err := cmd.CombinedOutput()
	if err != nil {
		errStr := strings.TrimSpace(string(out))
		if errStr == "" {
			errStr = err.Error()
		}
		return R("error: failed to kill PID " + strconv.Itoa(pid) + ": " + trunc(errStr, 200))
	}
	return R("killed PID " + strconv.Itoa(pid))
}
